Shrink PDF files with ghostscript before displaying (BL-3721)

StephenMcConnel opened the pull request:
What to do about ensuring ghostscript is installed on Windows is yet
to be determined or implemented.  There is a comment about what is
involved in the YouTrack issue.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/bloombooks/bloomdesktop/1699)
<!-- Reviewable:end -->

hatton commented:
Moved installer to new card, BL-4713

---

Review status: 0 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 56 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-KlvQJCOO_PTytNsD3oV:-KlvQJCOO_PTytNsD3oW:b-ejotoj) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ee81729814a370f373d548606ba56794e4649833/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L56)):*
> ```Smalltalk
> 			{
> 				if (_inputPdfPath != _outputPdfPath)
> 					RobustFile.Copy(_inputPdfPath, _outputPdfPath, true);
> ```

I don't know if the output  path here is reused... if so, we'd need to prepare for it to be locked.

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 80 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-KlvQVOlU8Q16ORsv09V:-KlvQVOlU8Q16ORsv09W:b-ttw8o5) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ee81729814a370f373d548606ba56794e4649833/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L80)):*
> ```Smalltalk
> 				if (_progress.ProgressStateResult != null && _progress.ProgressStateResult.ExceptionThatWasEncountered != null)
> 				{
> 					Console.WriteLine("Exception encountered fixing PDF - {0}", _progress.ProgressStateResult.ExceptionThatWasEncountered);
> ```

Instead of console, how about the report non fatal error toast system?

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 92 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-KlvRYSrYxsdjPb8ubSv:-KlvRYSrYxsdjPb8ubSw:bt1sj2q) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ee81729814a370f373d548606ba56794e4649833/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L92)):*
> ```Smalltalk
> 			if (!String.IsNullOrEmpty(exePath) && File.Exists(exePath))
> 			{
> 				var tempPdfFile = GetTemporaryPdfFilename();
> ```

Might as well use our normal tempfile pattern with a using(....)

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 173 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-KlvQstWUs9lCLy0ikJY:-KlvQstWUs9lCLy0ikJZ:bajv3wl) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ee81729814a370f373d548606ba56794e4649833/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L173)):*
> ```Smalltalk
> 
> 		private string SetArguments(string tempFile)
> 		{
> ```

This is more of "Get" then a "Set".

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 198 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-KlvR48_5X3B_CXxXZEF:-KlvR48_5X3B_CXxXZEG:b-8yehzu) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ee81729814a370f373d548606ba56794e4649833/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L198)):*
> ```Smalltalk
> 
> 		// Magic strings to match for initializing progress reporting.
> 		const string ProcessingPages = "Processing pages ";
> ```

we use a prefix "k" for constants.

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 199 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-KlvR94u-fEPgBzBc-At:-KlvR94u-fEPgBzBc-Au:b-6fxf7u) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ee81729814a370f373d548606ba56794e4649833/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L199)):*
> ```Smalltalk
> 		// Magic strings to match for initializing progress reporting.
> 		const string ProcessingPages = "Processing pages ";
> 		const string ThroughWithSpaces = " through ";
> ```

Will these strings that you match change on, say, a French system?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-:-KlvPHp1LNy1Xxl8_Qg0:b-5f40lz)*
<!-- Sent from Reviewable.io -->

StephenMcConnel commented:
Review status: 0 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 56 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-KlvQJCOO_PTytNsD3oV:-KlyFB-d55xuv6AnmaID:b-1nzosf) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ee81729814a370f373d548606ba56794e4649833/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L56)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

I don't know if the output  path here is reused... if so, we'd need to prepare for it to be locked.
</blockquote></details>

The calling code has tried hard to provide a filename that doesn't exist, by first trying to delete an existing file, and if that fails, adjusting the name and trying again until it gets a name that is free and clear.  (PublishModel.GetPdfPath())

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 80 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-KlvQVOlU8Q16ORsv09V:-KlyPu-8VpRI-n-nQKRm:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ee81729814a370f373d548606ba56794e4649833/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L80)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

Instead of console, how about the report non fatal error toast system?
</blockquote></details>

Done.

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 92 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-KlvRYSrYxsdjPb8ubSv:-Klynh61kKJoYyER1UIR:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ee81729814a370f373d548606ba56794e4649833/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L92)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

Might as well use our normal tempfile pattern with a using(....)
</blockquote></details>

Done.

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 173 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-KlvQstWUs9lCLy0ikJY:-KlyPsi0vFa8KKE1NqUw:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ee81729814a370f373d548606ba56794e4649833/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L173)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

This is more of "Get" then a "Set".
</blockquote></details>

Done.

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 198 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-KlvR48_5X3B_CXxXZEF:-KlyUDggTQmIE8XfQ3Ia:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ee81729814a370f373d548606ba56794e4649833/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L198)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

we use a prefix "k" for constants.
</blockquote></details>

Done.

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 199 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-KlvR94u-fEPgBzBc-At:-KlyPVQZdcXfMX24fc2H:brisrup) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ee81729814a370f373d548606ba56794e4649833/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L199)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

Will these strings that you match change on, say, a French system?
</blockquote></details>

I haven't seen any indication that ghostscript has tried to localize anything besides man pages (documentation) and maybe some sort of header files for printing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699)*
<!-- Sent from Reviewable.io -->

JohnThomson commented:
Nice piece of work. Various picky suggestions, some of which may be worth acting on.

---

Review status: 0 of 6 files reviewed at latest revision, 10 unresolved discussions.

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 50 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km2R7xqaXaTVL-cbV3n:-Km2R7xrl9jscxKJ03V-:b-flzrfi) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ad5724cd33eb880347b419f51906bd301931c2cb/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L50)):*
> ```Smalltalk
> 		}
> 
> 		public void CreateFixedPdfFile(string inputFile, string outputFile)
> ```

The name is a little vague, perhaps unavoidably since it may do two different things. I think a comment would be worthwhile.

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 54 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km2RNVvlYzIUX_Bp4sy:-Km2RNVvlYzIUX_Bp4sz:b6pcj3i) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ad5724cd33eb880347b419f51906bd301931c2cb/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L54)):*
> ```Smalltalk
> 			_inputPdfPath = inputFile;
> 			_outputPdfPath = outputFile;
> 			if (!_shrink && String.IsNullOrWhiteSpace(_rgbProfile) && String.IsNullOrWhiteSpace(_cmykProfile))
> ```

I gather it's an anomaly if _rgbProfile is null and _cmykProfile is not or vice versa? It feels odd to take the trouble to test both but not handle the mixed case somehow. Maybe at least assert that either both are null or neither is? Currently if I read correctly the result will be that neither the conversion nor the unchanged copy will take place.

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 82 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km2MHC8f_NvkI9MPF_D:-Km2MHC8f_NvkI9MPF_E:bka9gzx) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ad5724cd33eb880347b419f51906bd301931c2cb/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L82)):*
> ```Smalltalk
> 				{
> 					var shortMsg = String.Format("Exception encountered fixing PDF: {0}", _progress.ProgressStateResult.ExceptionThatWasEncountered);
> 					NonFatalProblem.Report(ModalIf.None, PassiveIf.Beta, shortMsg, null, _progress.ProgressStateResult.ExceptionThatWasEncountered);
> ```

We really don't want to give any indication that there was a problem in a release build or return any indication to the caller? What will Bloom do if the expected output file is missing or corrupt? But then, you do catch most problems and revert to copying the file inside DoFixPdfWork...will we typically get both toasts? Or one of them will overlay the other? Seems like we just want one or the other of these...
Another thought...are these messages likely enough to be seen to deserve localization?

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 109 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km2NPQ0YpZsRqKJlgcG:-Km2NPQ0YpZsRqKJlgcH:b-d35dys) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ad5724cd33eb880347b419f51906bd301931c2cb/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L109)):*
> ```Smalltalk
> 						return;
> 					}
> 					// If the process actually made the file larger, ignore the result.
> ```

Surely we only want to suppress a bigger output if we were trying to shrink it, not if we're converting to CMYK?

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 120 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km2NjngeNlMZkD4tbpX:-Km2NjngeNlMZkD4tbpY:b-22pyue) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ad5724cd33eb880347b419f51906bd301931c2cb/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L120)):*
> ```Smalltalk
> 			else
> 			{
> 				// REVIEW: ask user to download and install ghostscript?  how often?  only once?  once per run of Bloom and saving PDF?
> ```

It's one thing to quietly not shrink the file, but if the user asked for CMYK and we can't provide it, seems we need to inform them somehow. Maybe just a TODO at this point?

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 130 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km2Nw9tuy7OG7W4GMRi:-Km2Nw9tuy7OG7W4GMRj:b-wdto1f) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ad5724cd33eb880347b419f51906bd301931c2cb/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L130)):*
> ```Smalltalk
> 		{
> 			var newTempFile = Path.GetTempFileName();
> 			File.Delete(newTempFile);
> ```

It seems pointless to delete a file that isn't the path we're going to use.
I have a vague recollection that our TempFile util class as a method like CreateWithExtension that might be helpful, if you can arrange to dispose it at an appropriate time.

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 164 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km2P47ixpJmMuXXqs-W:-Km2P47ixpJmMuXXqs-X:b-a6ztpx) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ad5724cd33eb880347b419f51906bd301931c2cb/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L164)):*
> ```Smalltalk
> 			if (!Directory.Exists(baseDir))
> 				return null;
> 			foreach (var versionDir in Directory.GetDirectories(baseDir))
> ```

Do we have any preference for later versions over earlier ones if more than one is installed? If so is there any easy way to sort them?
Is there any limit to how early a version is useful at all?

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 179 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km2PPNHGVGKjxKT8p8i:-Km2PPNHGVGKjxKT8p8j:b-uas1ho) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ad5724cd33eb880347b419f51906bd301931c2cb/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L179)):*
> ```Smalltalk
> 		{
> 			var bldr = new StringBuilder();
> 			bldr.Append("-sDEVICE=pdfwrite -dBATCH -dNOPAUSE -dCompatibilityLevel=1.7");	// REVIEW: is this the right compatibility level?
> ```

Yes, is it???

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 192 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km2Pf0hzgwhcZ5d10lO:-Km2Pf0hzgwhcZ5d10lP:benxxp3) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ad5724cd33eb880347b419f51906bd301931c2cb/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L192)):*
> ```Smalltalk
> 			if (!String.IsNullOrEmpty(_rgbProfile) && !String.IsNullOrEmpty(_cmykProfile))
> 			{
> 				// TODO: handle _rgbProfile and _cmykProfile  (BL-4457)
> ```

I gather this is deliberately not in scope for this PR?

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 194 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km2PsOoSUovgWnxYYNg:-Km2PsOoSUovgWnxYYNh:b-3t6knj) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ad5724cd33eb880347b419f51906bd301931c2cb/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L194)):*
> ```Smalltalk
> 				// TODO: handle _rgbProfile and _cmykProfile  (BL-4457)
> 			}
> 			bldr.AppendFormat(" -sOutputFile=\"{0}\" \"{1}\"", tempFile, _inputPdfPath);
> ```

There's an elegant new way to do this in C# 6
bldr.Append($" -sOutputFile=\"{tempFile}\" \"{_inputPdfPath}\")  (The $ in front of the string triggers the automatic insertion of variables in braces)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-:-Km2TPq8LAARztT0Pjpe:b-i7lauv)*
<!-- Sent from Reviewable.io -->

JohnThomson commented:
Reviewed 5 of 6 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699)*
<!-- Sent from Reviewable.io -->

StephenMcConnel commented:
Review status: 2 of 6 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 50 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km2R7xqaXaTVL-cbV3n:-Km39OM3jRP3tGl9auMH:b-iy8c1q) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ad5724cd33eb880347b419f51906bd301931c2cb/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L50)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

The name is a little vague, perhaps unavoidably since it may do two different things. I think a comment would be worthwhile.
</blockquote></details>

I renamed to something even vaguer (if possible), but added an method comment.

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 54 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km2RNVvlYzIUX_Bp4sy:-Km3AIO9ohPXFyLurKic:b-l1k7hl) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ad5724cd33eb880347b419f51906bd301931c2cb/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L54)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

I gather it's an anomaly if _rgbProfile is null and _cmykProfile is not or vice versa? It feels odd to take the trouble to test both but not handle the mixed case somehow. Maybe at least assert that either both are null or neither is? Currently if I read correctly the result will be that neither the conversion nor the unchanged copy will take place.
</blockquote></details>

Added an assert in the constructor, and a couple of comments to clarify the situation.

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 82 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km2MHC8f_NvkI9MPF_D:-Km3C-JOS1J0Y6qSPdJO:bjlfg2v) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ad5724cd33eb880347b419f51906bd301931c2cb/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L82)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

We really don't want to give any indication that there was a problem in a release build or return any indication to the caller? What will Bloom do if the expected output file is missing or corrupt? But then, you do catch most problems and revert to copying the file inside DoFixPdfWork...will we typically get both toasts? Or one of them will overlay the other? Seems like we just want one or the other of these...
Another thought...are these messages likely enough to be seen to deserve localization?
</blockquote></details>

I changed it to PassiveIf.All, and removed the problem report inside the DoWork method.  I also made the short message localizable and more specific, and added a long message (which we don't normally localize since it goes into an error report presumably read by a developer).

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 109 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km2NPQ0YpZsRqKJlgcG:-Km3FjUwXyS-tA_lhA7s:biha3b4) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ad5724cd33eb880347b419f51906bd301931c2cb/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L109)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Surely we only want to suppress a bigger output if we were trying to shrink it, not if we're converting to CMYK?
</blockquote></details>

Good catch.  Added check for color change as well as file size.

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 120 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km2NjngeNlMZkD4tbpX:-Km3Fsk-rSCDxvoC4xwE:blmpnk3) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ad5724cd33eb880347b419f51906bd301931c2cb/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L120)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

It's one thing to quietly not shrink the file, but if the user asked for CMYK and we can't provide it, seems we need to inform them somehow. Maybe just a TODO at this point?
</blockquote></details>

This will probably be handled in the initial menu click code.  The UI design at the moment doesn't allow the file compression to be optional, but there is a menu item specified for selecting the color change.  That's where we'll interact with the user about things like the availability of ghostscript.  In practice, the Linux package depends on ghostscript so it should always be available there, and JohnH wants to add ghostscript files to our installer so that it will always be available there as well.  So he's answered my review question I guess.  Maybe I should remove it.  :-)

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 130 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km2Nw9tuy7OG7W4GMRi:-Km3H9b0GxRXze-i_pc-:bdti2ln) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ad5724cd33eb880347b419f51906bd301931c2cb/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L130)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

It seems pointless to delete a file that isn't the path we're going to use.
I have a vague recollection that our TempFile util class as a method like CreateWithExtension that might be helpful, if you can arrange to dispose it at an appropriate time.
</blockquote></details>

This was a leftover method that I forgot to delete after adjusting the code to use TempFile.CreateWithExtension in a using block.

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 164 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km2P47ixpJmMuXXqs-W:-Km3HMF8jxOhrq7MitUJ:bmz6adx) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ad5724cd33eb880347b419f51906bd301931c2cb/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L164)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Do we have any preference for later versions over earlier ones if more than one is installed? If so is there any easy way to sort them?
Is there any limit to how early a version is useful at all?
</blockquote></details>

After the installer gets updated to include ghostscript, this code needs to be adjusted to look there first (and maybe only).

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 179 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km2PPNHGVGKjxKT8p8i:-Km3LfwAITah5LA9eWL7:b-9x5xqy) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ad5724cd33eb880347b419f51906bd301931c2cb/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L179)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Yes, is it???
</blockquote></details>

Good question.  1.7 dates back to 2008 (Acrobat 8.0) so it should be safe one would think.  I've added some notes about compatibility levels and versions of Acrobat from

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 192 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km2Pf0hzgwhcZ5d10lO:-Km3HanCddtGSA1-ccLK:b-86u2cp) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ad5724cd33eb880347b419f51906bd301931c2cb/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L192)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

I gather this is deliberately not in scope for this PR?
</blockquote></details>

That is correct.  This PR handles BL-3721.  The color adjustment is BL-4457, but it builds on this code so I did some placeholders as part of this work.

---

*[src/BloomExe/Publish/FixPdfUsingGhostscript.cs, line 194 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km2PsOoSUovgWnxYYNg:-Km3MnY_6bKs58sCoE1z:b-73wfgy) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/ad5724cd33eb880347b419f51906bd301931c2cb/src/BloomExe/Publish/FixPdfUsingGhostscript.cs#L194)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

There's an elegant new way to do this in C# 6
bldr.Append($" -sOutputFile=\"{tempFile}\" \"{_inputPdfPath}\")  (The $ in front of the string triggers the automatic insertion of variables in braces)
</blockquote></details>

I guess it's worth a try.  :-)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699)*
<!-- Sent from Reviewable.io -->

JohnThomson commented:
Looking pretty good now. I think I caught one new problem, and one more take-or-leave comment.

---

Review status: 2 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/BloomExe/Publish/ProcessPdfWithGhostscript.cs, line 110 at r3](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km3Qj7AX1xxrMtWlpJ8:-Km3Qj7AX1xxrMtWlpJ9:b-3dix77) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/dae6110cdf834ba944f0448f6c6938cf8c847f19/src/BloomExe/Publish/ProcessPdfWithGhostscript.cs#L110)):*
> ```Smalltalk
> 					}
> 					var longMsg = String.Format("Exception encountered processing the PDF file: {0}", _progress.ProgressStateResult.ExceptionThatWasEncountered);
> 					NonFatalProblem.Report(ModalIf.None, PassiveIf.All, null, longMsg, _progress.ProgressStateResult.ExceptionThatWasEncountered);
> ```

I think you meant to pass shortMsg somewhere here?

---

*[src/BloomExe/Publish/ProcessPdfWithGhostscript.cs, line 138 at r3](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km3R6fas9sJAx4ZAGT5:-Km3R6fas9sJAx4ZAGT6:b-23lw78) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/dae6110cdf834ba944f0448f6c6938cf8c847f19/src/BloomExe/Publish/ProcessPdfWithGhostscript.cs#L138)):*
> ```Smalltalk
> 					var oldInfo = new FileInfo(_inputPdfPath);
> 					var newInfo = new FileInfo(tempPdfFile.Path);
> 					if (newInfo.Length < oldInfo.Length || (!String.IsNullOrEmpty(_rgbProfile) && !String.IsNullOrEmpty(_cmykProfile)))
> ```

Since we're now asserting that either both are null or neither is, I think it would reduce complexity usefully to just test one of them here and in similar places.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-:-Km3Rl-XLAx1Qq5qjH6M:b3ssmhy)*
<!-- Sent from Reviewable.io -->

StephenMcConnel commented:
Review status: 2 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/BloomExe/Publish/ProcessPdfWithGhostscript.cs, line 110 at r3](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km3Qj7AX1xxrMtWlpJ8:-Km3TPB30OG2zFrKPVaH:b-sdzbw) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/dae6110cdf834ba944f0448f6c6938cf8c847f19/src/BloomExe/Publish/ProcessPdfWithGhostscript.cs#L110)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

I think you meant to pass shortMsg somewhere here?
</blockquote></details>

Done.  oops.

---

*[src/BloomExe/Publish/ProcessPdfWithGhostscript.cs, line 138 at r3](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-Km3R6fas9sJAx4ZAGT5:-Km3Ub6WO2ir-2qKUhv5:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/dae6110cdf834ba944f0448f6c6938cf8c847f19/src/BloomExe/Publish/ProcessPdfWithGhostscript.cs#L138)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Since we're now asserting that either both are null or neither is, I think it would reduce complexity usefully to just test one of them here and in similar places.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699)*
<!-- Sent from Reviewable.io -->

hatton commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 7 files at r1, 1 of 2 files at r2.
Review status: 2 of 6 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-:-Km4Ry38YW7OmuojEN2d:bnfp4nl)*
<!-- Sent from Reviewable.io -->

hatton commented:
Please make the compressing step show up in the same progress dialog as the pdf creation.

hatton commented:
Please change the compression statement to parallel the creation statement. "Creating PDF...", "Compressing PDF...".

StephenMcConnel commented:
The problem with using the same progress dialog is that the creation progress dialog is internal to GeckofxHtmlToPdf (aka BloomPdfMaker) and the compression work is being done in Bloom proper.  If you really want the same progress dialog to be used, I'll have to move the ghostscript processing code back from Bloom to GeckofxPdfToHtml.  I hesitate to do this for three reasons, one important and two probably not. 

1) GeckofxPdfToHtml (GPTH) does not use any libpalaso code on purpose to minimize its dependencies but the ghostscript processing code does use some libpalaso (or Bloom?) code such as the Toast notification.
2) When the code was over in GPTH, I was never able to get the progress report to actually respond to the ghostscript progress.  The right calls were being made, I even checked for and used Invoke if needed, but the progress dialog never responded or changed its appearance. 
3) I've already moved the code from GPTH to Bloom, and it will take some work to move it back.  (See reason 2 as adding to this work.)

I suppose an alternative would be to rewrite GTPH to pump out messages to the console as a console app instead, and trap and handle those like I'm trapping and interpreting the progress reports from ghostscript.  But that's a rather major change in its own right, and would also more closely couple the version of GHTP to the version of Bloom.

Changing the compression statement, on the other hand, is rather trivial and I can certainly do that.  :-)

---

Review status: 2 of 6 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-:-Km75jy4BQp29JpRRf5D:b-4jgy3f)*
<!-- Sent from Reviewable.io -->

JohnThomson commented:
A somewhat lower-cost option would be to pass the PDFmaker tool arguments telling it that its progress should finish at some typical percentage...say 80...and where to locate the progress dialog. Then Bloom could put its own dialog in the same place and advance it from 80 to 100 while the compression goes on. If it's worth it, at some further cost we could probably make the two progress dialogs more similar in appearance. Then there's just a brief glitch where the dialog disappears and reappears...maybe something could be done so that Bloom displays the dialog sooner, even before it's ready to launch ghostscript?

---

Review status: 2 of 6 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-:-Km7xqM-0JIZfu9v89x4:b-2zo4x4)*
<!-- Sent from Reviewable.io -->

hatton commented:
Steve & John thanks for thinking up options. It seems to me that the console option is actually just 2-3 hours work, and has the advantage that we could then localize all the messages coming out of the pdf creation process. Steve is that true?

StephenMcConnel commented:
I've implemented the console logging/reading option.  It took longer than 2-3 hours, partly because of a Mono runtime bug that kept the original progress dialog from being hidden.  It took me a while to work around that problem.  Then getting the L10N aspects of converting the output took a bit longer as well.  But I did have the basics working within 3 hours.  :-)

---

Review status: 2 of 8 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1699#-:-KmDiPI6GAK9r2npmb6T:b-dbeza4)*
<!-- Sent from Reviewable.io -->
